### ISSUE TITLE
Unable to set a structure member's custom type, it always is 'int'

### ISSUE THREAD

**Issue by arizvisa** [arizvisa/ida-minsc#62]
@arizvisa hi, I have to come and ask you questions, I don't know what went wrong, I can't set the member's custom type , it awaly is 'int':

```
import idaapi
import idc


st1 =structure.by_name('TTDBTextInfoGroup')
struct_name = 'TTDBTextInfoGroup*'
f1 = st1.members.by(offset=0)

print f1.typeinfo.dstr()
ordinal = f1.typeinfo.get_ordinal()
f1.typeinfo.set_numbered_type(idaapi.cvar.idati,ordinal,BTF_STRUCT, struct_name)
print f1.typeinfo.dstr()
```

_Originally posted by @gool123456 in https://github.com/arizvisa/ida-minsc/issues/61#issuecomment-664488530_

**Comment by arizvisa** [comment]
Can you print out the structure st1 to me? There might not be enough space in the field to promote it to an int.

This is actually related to issue #45. IDA doesn't return an error or anything, so to solve issue #45, we need to verify that the field was properly assigned.

_Originally posted by @arizvisa in https://github.com/arizvisa/ida-minsc/issues/61#issuecomment-664504271_

**Comment by arizvisa** [comment]
Also, are you trying to change the type of the member at offset 0 to a structure?

If so, then you'll want to do:
```
st = struc.by('mystructure')
m = st.members.by(offset=0)
m.type = struc.by('structuremember')
```

_Originally posted by @arizvisa in https://github.com/arizvisa/ida-minsc/issues/61#issuecomment-664505803_

**Comment by arizvisa** [comment]
You can also treat `member.typeinfo` as a string, meaning you can do:
```
m.typeinfo = 'char*'
```

_Originally posted by @arizvisa in https://github.com/arizvisa/ida-minsc/issues/61#issuecomment-664506222_

**Comment by gool123456** [comment]
@ Thanks , but ida 6.8 unfind the idaapi.get_idati , The error happend at structure.py -> typeinfo line 1619 :

I try fix it :

@typeinfo.setter
@utils.string.decorate_arguments('info')
def typeinfo(self, info):
    '''Set the type info of the member to `info`.'''
    if idaapi.__version__ < 7.0:
        **til, ti = idaapi.cvar.idati , idaapi.tinfo_t()   #added**
    else:
        til, ti = idaapi.get_idati(), idaapi.tinfo_t(),



But then the error occurred :

```
Script Default snippet error: Traceback (most recent call last):
  File "<string>", line 7, in <module>
  File "C:\Users\Hskyre\AppData\Roaming\Hex-Rays\IDA Pro\base\structure.py", line 1615, in typeinfo
    @typeinfo.setter
  File "C:\Users\Hskyre\AppData\Roaming\Hex-Rays\IDA Pro\base\_utils.py", line 988, in wrapper
    return F(*res, **kwds)
  File "C:\Users\Hskyre\AppData\Roaming\Hex-Rays\IDA Pro\base\structure.py", line 1634, in typeinfo
    res = idaapi.parse_decl(ti, til, terminated, idaapi.PT_SIL)
  File "D:\tools\IDA_Pro_v6.8_and_Hex-Rays_Decompiler_(ARM,x64,x86)_Hskyre\python\idaapi.py", line 35719, in parse_decl
    return _idaapi.parse_decl(*args)
TypeError: parse_decl expected 6 arguments, got 4
```

**Comment by arizvisa** [comment]
Ahh. Okay. Seems like IDA 6.8's API is completely different from IDA 7.x's. Lemme see what IDA 6.8 expects you to do here and create a PR.

**Comment by arizvisa** [comment]
@gool123456, can you try out PR #62 and see if that remedies the issue that you mentioned?

**Comment by arizvisa** [comment]
Ermm. I mean PR #63.

**Comment by gool123456** [comment]
@arizvisa Received, please forgive me for not receiving your message in time. i, trying !

**Comment by gool123456** [comment]
@arizvisa Show Error :

```
Script Default snippet error: Traceback (most recent call last):
  File "<string>", line 7, in <module>
  File "C:\Users\Hskyre\AppData\Roaming\Hex-Rays\IDA Pro\base\structure.py", line 1615, in typeinfo
    @typeinfo.setter
  File "C:\Users\Hskyre\AppData\Roaming\Hex-Rays\IDA Pro\base\_utils.py", line 988, in wrapper
    return F(*res, **kwds)
  File "C:\Users\Hskyre\AppData\Roaming\Hex-Rays\IDA Pro\base\structure.py", line 1634, in typeinfo
    res = idaapi.parse_decl2(til, terminated, ti, idaapi.PT_SIL) if idaapi.__version__ < 7.0 else idaapi.parse_decl(ti, til, terminated, idaapi.PT_SIL)
  File "D:\tools\IDA_Pro_v6.8_and_Hex-Rays_Decompiler_(ARM,x64,x86)_Hskyre\python\idaapi.py", line 32392, in parse_decl2
    return _idaapi.parse_decl2(*args)
TypeError: parse_decl2 expected 5 arguments, got 4

```

**Comment by arizvisa** [comment]
Wow. That is so funny. According to https://www.hex-rays.com/products/ida/support/ida74_idapython_no_bc695_porting_guide.shtml, the function prototype or signature for `idaapi.parse_decl2` is

```
Earlier IDA: ida_typeinf.parse_decl2(til, decl, tif, flags)
Newer IDA: ida_typeinf.parse_decl(tif, til, decl, flags)
```

**Comment by arizvisa** [comment]
Okay. I don't know if it'll work because I'm still downloading IDA 6.8 to reproduce (my internet is very slow), but the sdk says that the prototype is.

```
idaman bool ida_export | parse_decl2 (til_t *til, const char *decl, qstring *name, tinfo_t *tif, int flags)
```

To fix it, I passed it as 
```
+        return None if idaapi.parse_decl2(til, terminated, None, ti, idaapi.PT_SIL) is None else ti
```

Can you try PR #63 again?

**Comment by gool123456** [comment]
@arizvisa Then, PR#63 it showed these errors:

```
Script Default snippet error: Traceback (most recent call last):
  File "<string>", line 6, in <module>
  File "C:\Users\Hskyre\AppData\Roaming\Hex-Rays\IDA Pro\base\structure.py", line 1627, in typeinfo
    res = idaapi.set_member_tinfo(self.parent.ptr, self.ptr, self.ptr.get_soff(), ti, 0)
  File "D:\tools\IDA_Pro_v6.8_and_Hex-Rays_Decompiler_(ARM,x64,x86)_Hskyre\python\idaapi.py", line 53264, in set_member_tinfo
    return _idaapi.set_member_tinfo(*args)
TypeError: set_member_tinfo expected 7 arguments, got 5
```

**Comment by arizvisa** [comment]
Okay. Try PR #63 again. I also cleaned up the typeinfo so that it's properly emitted when displaying the structure member or listing them with `struc.list()`.

**Comment by arizvisa** [comment]
Unfortunately I found another (visual) bug related to IDA not using `sys.displayhook` when writing its output. So I'll resolve that bug in another issue.

**Comment by gool123456** [comment]
Nice ! everything is normal. There are just a lot of warnings like this, I don’t know if it helps you? :

ING:root:internal.interface.priorityhook.callback(8592792, False) : 
ReadOrWriteError: internal.comment.contents.set_address(0x831d98, 0) : Unable to write name to address 0x831d98.
WARNING:root:internal.interface.priorityhook.callback(8592792, False) : 
......

**Comment by arizvisa** [comment]
Are the structure member issues fixed by that PR?

**Comment by arizvisa** [comment]
Let me know if it does so I can close this issue, merge it, and then continue to work on issue #61.

**Comment by arizvisa** [comment]
Promoted this issue to a bug as it documents issues discovered with setting structure member types.

**Comment by arizvisa** [comment]
@gool123456, I'm pretty sure PR #63 fixes this issue. Can you confirm when you wake up and get a chance?

**Comment by gool123456** [comment]
@arizvisa No problem, everything is ok! :) Thank you .

**Comment by gool123456** [comment]
HAHA , There are a lot of RPs, and sometimes I can't find a suitable place to reply to you.

**Comment by arizvisa** [comment]
Lol. Typically I try to separate the conversation as the following.

Issues: Used for identifying the bug, and how to reproduce it. Also includes questions related to the bug or some feature.
Pull Requests (PRs): Used for fixing the bug, discussion of how to fix it, perhaps issues related to fixing the bug.

**Comment by arizvisa** [comment]
Closing this ticket as it has been fixed by PR #63, and confirmed by the OP.
